### ISSUE TITLE
Cannot build upon fresh clone

### ISSUE THREAD

**Issue by ockham** [woocommerce/woocommerce-gutenberg-products-block#10166]
## Describe the bug

Using a fresh clone of the repo, running the build aborts with failures.

## To reproduce

1. Check out a fresh clone of the repo.
2. `npm i`
3. `composer install`
4. `npm run build`

## Expected behavior

The build fails with a long list of `Module parse failed: Unexpected token` errors (furthermore stating that `You may need an additional loader to handle the result of these loaders`). The following are the three errors at the top (but the list goes on much longer):

```
    ERROR in ./assets/js/middleware/store-api-nonce.js 12:29
    Module parse failed: Unexpected token (12:29)
    File was processed with these loaders:
     * ./node_modules/babel-loader/lib/index.js
    You may need an additional loader to handle the result of these loaders.
    |   const storedNonceValue = window.localStorage.getItem('storeApiNonce');
    |   const storedNonce = storedNonceValue ? JSON.parse(storedNonceValue) : {};
    >   currentNonce = storedNonce?.nonce || '';
    |   currentTimestamp = storedNonce?.timestamp || 0;
    | } catch {
    
    ERROR in ./assets/js/base/components/block-error-boundary/index.tsx 13:8
    Module parse failed: Unexpected token (13:8)
    File was processed with these loaders:
     * ./node_modules/babel-loader/lib/index.js
    You may need an additional loader to handle the result of these loaders.
    | import './style.scss';
    | class BlockErrorBoundary extends Component {
    >   state = {
    |     errorMessage: '',
    |     hasError: false
    
    ERROR in ./assets/js/settings/blocks/constants.ts 16:41
    Module parse failed: Unexpected token (16:41)
    File was processed with these loaders:
     * ./node_modules/babel-loader/lib/index.js
    You may need an additional loader to handle the result of these loaders.
    | export const WC_BLOCKS_BUILD_URL = blocksConfig.pluginUrl + 'build/';
    | export const WC_BLOCKS_PHASE = blocksConfig.buildPhase;
    > export const SHOP_URL = STORE_PAGES.shop?.permalink;
    | export const CHECKOUT_PAGE_ID = STORE_PAGES.checkout.id;
    | export const CHECKOUT_URL = STORE_PAGES.checkout.permalink;
```

## Environment

- Node 16.19.1
- npm 8.19.3

## Additional context

Looks like it's complaining about ES6+ constructs? 🤔

**Comment by nielslange** [comment]
Hello @ockham, 

Thanks for reaching out! Before running the commends you mentioned in the description, could you run the following command:
```sh
nvm use
```

In the description, you mentioned that you are using the following environment:
- Node 16.19.1
- npm 8.19.3

The command, I mentioned, ensures that you are running this environment instead:
- Node 16.15.0
- npm 8.5.5

The corresponding environment definition can be found in https://github.com/woocommerce/woocommerce-blocks/blob/trunk/.nvmrc. In case you are unfamiliar with `nvm`, it's the Node Version Manager and on https://github.com/nvm-sh/nvm you can find further information such as how to install it.

I hope this helps to solve the problem that you are currently facing.

**Comment by ockham** [comment]
@nielslange Wow, using node 16.15.0 and npm 8.5.5 actually did the trick -- thanks a lot!

(FWIW, I'm using `n` (rather than `nvm`) as my node version management tool, which is why the `.nvmrc` config wasn't automatically picked up.)

I feel like this could use at least a bit more documentation in the README, or ideally a fix that would make the build compatible with other 16.x versions of node 😅

**Comment by ockham** [comment]
Looping in webpack experts-in-residence @jsnajdr and @tyxla -- just in case you've come across any weirdnesses around ES6+ related parse errors that appeared in certain 16.x versions of node, but were absent in 16.15 😅  (Feel free to ignore otherwise.)

**Comment by nielslange** [comment]
> @nielslange Wow, using node 16.15.0 and npm 8.5.5 actually did the trick -- thanks a lot!

Nice! Glad to hear that, @ockham! 🙌

> (FWIW, I'm using n (rather than nvm) as my node version management tool, which is why the .nvmrc config wasn't automatically picked up.)
> 
> I feel like this could use at least a bit more documentation in the README, or ideally a fix that would make the build compatible with other 16.x versions of node 😅

Thanks for sharing the additional information. Occasionally, we bump the supported node version. The problem is that it requires a lot of testing to ensure that everything remains working. You can imagine that we're careful when it comes to bumping the supported node version. That said, we should update our documentation. Currently, it says _"Make sure you're using Node.js v16."_, which is not 100% correct. I'll create an issue for that.

**Comment by ockham** [comment]
> Occasionally, we bump the supported node version. The problem is that it requires a lot of testing to ensure that everything remains working. You can imagine that we're careful when it comes to bumping the supported node version. 

Yeah, that's understandable (and somewhat similar to how it's handled e.g. in the Gutenberg repo). FWIW, what stood out to me as a bit surprising is that really only Node 16.15 seemed to work out of the entire 16.x series -- I had tried 16.14 and 16.17, and neither worked. This appeared to me a bit fragile, and potentially related to some underlying bug in the build system that just happened to work with 16.15 😅 I guess that ideally, the build system should be made a bit more resilient; I see that @Aljullu has been working on a potential fix in #9940 😊 

> That said, we should update our documentation. Currently, it says "Make sure you're using Node.js v16.", which is not 100% correct. I'll create an issue for that.

Thank you!

**Comment by jsnajdr** [comment]
The `woocommerce-blocks` repo needs to upgrade its webpack from 4.44.2, which is 3 years old, to a more recent 5.x version.

The "Module parse failed" errors happen on the optional chaining and class properties syntax. Webpack uses the `acorn` parser to parse JS modules, and the old 4.x version of webpack uses a version of `acorn` that is too old to support this syntax. That's why it fails.

The optional chaining and class properties syntax used to be transpiled by Babel, but today they are not transpiled any more. The `@wordpress/browserslist-config` package contains definitions of browsers supported by Gutenberg: the supported browsers can be summarized as "last 2 versions of major browsers + browsers with more than 1% usage". Then the `@babel/preset-env` plugin automatically chooses what to transpile based on the list of supported browsers. As time goes, all of the supported browsers now support optional chaining and class properties. Also, in https://github.com/wordpress/gutenberg/pull/50994, I merged a Babel config optimization that leads to optional chaining being mostly untranspiled.

The reason why build fails with Node 16.15 and not with 16.19 is not because the versions would be that much different internally. It's because of npm: the version 8.19.3 that ships with Node 16.19 doesn't like the `woocommerce-blocks` repo's lockfile. `npm ci` fails with: 
```
`npm ci` can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
```
and you need to do a full `npm install` to get a `node_modules` tree. But that installs different versions of many packages. Apparently somehow a newer version of `acorn` or something else sneaks in somewhere, and suddenly the new JS syntax _can_ be parsed.

The changes that @Aljullu is proposing in #9940 are enforcing transpilation of the optional chaining and class properties again, after the combination of `@wordpress/browserslist-config` and `@babel/preset-env` stopped transpiling them automatically. That fixes the problem with old `acorn` being unable to parse the transpiled syntax, but it goes in the wrong direction. Having less transpiled code is actually desirable. We just need to upgrade the rest of the toolchain to be able to parse the new syntax.

**Comment by Aljullu** [comment]
Thanks so much for sneaking in @jsnajdr! We do have the goal to update to Webpack 5 (see in progress PR: https://github.com/woocommerce/woocommerce-blocks/pull/8013), but we didn't have time to finish it up yet.

> The changes that @Aljullu is proposing in #9940 are enforcing transpilation of the optional chaining and class properties again, after the combination of `@wordpress/browserslist-config` and `@babel/preset-env` stopped transpiling them automatically. That fixes the problem with old `acorn` being unable to parse the transpiled syntax, but it goes in the wrong direction. Having less transpiled code is actually desirable. We just need to upgrade the rest of the toolchain to be able to parse the new syntax.

That makes a lot of sense, thanks! :bow: 

---

@tjcafferkey @danielwrobert given the discussion above and @jsnajdr insights, it's probably a good idea for me to stop working on #9940 and instead prioritize our work on #8013. Do you agree? Also, cc @thealexandrelara as you were working on the Webpack update PR.

tl;dr: in several circumstances our builds fail, that's currently blocking renovatebot PRs and it's causing issues when building with certain node versions. In https://github.com/woocommerce/woocommerce-blocks/pull/9940 I was researching an interim solution, but it has proven time-consuming (I couldn't solve all issues yet) and as @jsnajdr mentioned, it's probably going in the wrong direction.

**Comment by tjcafferkey** [comment]
@Aljullu thanks for the ping, and also @jsnajdr for the write up!

I agree, I think there's more value in https://github.com/woocommerce/woocommerce-blocks/pull/8013

**Comment by thealexandrelara** [comment]
> @tjcafferkey @danielwrobert given the discussion above and @jsnajdr insights, it's probably a good idea for me to stop working on https://github.com/woocommerce/woocommerce-blocks/pull/9940 and instead prioritize our work on https://github.com/woocommerce/woocommerce-blocks/pull/8013. Do you agree? Also, cc @thealexandrelara as you were working on the Webpack update PR.

Thanks for the ping, @Aljullu. That makes sense. I think the PR is almost ready. The last few times I started working on it again, I faced some issues with the e2e tests or the pipeline. I wasn't confident whether the changes were breaking something or not because some of the tests were failing. I started investigating, and it seemed related to the translations, but I didn't delve deeper because I had to prioritize other tasks, I apologize for not giving enough attention to that PR in the past few weeks.

To prepare the terrain for you to work on it, @Aljullu, I just merged trunk and resolved some conflicts that appeared. Some e2e tests are still failing, so it's something that needs attention, let me know if there is anything else I can assist you with 🙌

**Comment by Aljullu** [comment]
> Thanks for the ping, @Aljullu. That makes sense. I think the PR is almost ready. The last few times I started working on it again, I faced some issues with the e2e tests or the pipeline. I wasn't confident whether the changes were breaking something or not because some of the tests were failing. I started investigating, and it seemed related to the translations, but I didn't delve deeper because I had to prioritize other tasks, I apologize for not giving enough attention to that PR in the past few weeks.

No need to apologize! :slightly_smiling_face:  I know there are other tasks which had more priority these last months. I just wanted to bring attention to this task as it's now blocking dependabot PRs, so I think its priority should be increased and maybe include this work in the next Cycle.

**Comment by dinhtungdu** [comment]
@ockham We upgraded `webpack` to version 5. I built successfully a fresh clone using your node version mentioned in the first post. I'm closing this one but feel free to reopen it if you still have issues.
